### DISCUSSION TITLE
Why is HealthCheck memoized?

### DISCUSSION THREAD

**Discussion by parthik19** [gvolpe/pfps-shopping-cart#572] [General]
The `status` field in `HealthCheck[F]` is `val` and not `def` on line 46 of `Healthcheck.scala`. Doesn't this mean that whenever status is called, `healthCheck.status` will return whatever is cached from the _first_ call of `status`? Shouldn't it recompute the status since a successful status at one point doesn't mean the service will be healthy at every point afterwards?

**Comment by gvolpe** [comment]
Hi @parthik19, it is not memoized, that is the power of [referential](https://livebook.manning.com/concept/scala/referential-transparency) [transparency](https://www.youtube.com/watch?v=X-cEGEJMx_4) :)

This is a key property of functional programming. It allows us to treat [program as values](https://systemfw.org/posts/programs-as-values-I.html); we can inline a value such as one returning `F[RedisStatus]` without changing the meaning of the resulting program. 

Have a look at those links, and I also recommend to play around with similar code that uses `Console[F].println` instead. You'll see that the different calls to `status` are evaluated per every call. There's no memoization whatsoever :)

**Reply by parthik19** [reply]
Thank you for the explanation! Allow me to ask another question - 

_Hypothetically_, suppose `HealthCheck[F]` defined its `status` field as `Sync[F].pure(/* compute AppStatus here */)`

Then, it would memoized, correct?

**Reply by BalmungSan** [reply]
Hi @parthik19 yeah you are correct, in that case the result would be _"memoized"_, since it would even be evaluated at the moment of creation of the program / value. That is why `pure` only make sense for plain values which could be inlined without affecting the RT of the whole expression.

**Reply by parthik19** [reply]
Okay makes sense, thanks!
